### PR TITLE
AEA-1385 fipa end2end test for python and go

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

fipa end2end test for python and go

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes and CI passes too
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by solarw** [comment]
Test file will be moved to another place after updated with go code. placed in `tests` just for CI check.

File: tests/test_fipa_end2end.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+    addr: str
+
+    def setup(self) -> None:
+        """Set up behaviour."""
+        self.is_started = False
+        self.was_called = False
+
+    def act(self) -> None:
+        """Make an action."""
+        self.dialogues = self.context.shared_state["dialogues"]
```
Comment (DavidMinarsch) [inline tests/test_fipa_end2end.py:114]: Dialogues should not be stored in shared_state. Just create in init! The shared_state is for cross-skill usage.

File: tests/test_fipa_end2end.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+    def setup(self) -> None:
+        """Set up behaviour."""
+        self.got_proposal = False
+
+    def teardown(self) -> None:
+        """Tear down handler."""
+
+    def handle(self, message) -> None:
+        """Handle an evelope."""
+        self.dialogues = self.context.shared_state["dialogues"]
```
Comment (DavidMinarsch) [inline tests/test_fipa_end2end.py:146]: see above

File: tests/test_fipa_end2end.py [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            skill_context=skill_context,
+            handlers={"handler": handler},
+            behaviours={"behaviour": behaviour},
+        )
+        builder.add_component_instance(test_skill)
+        with cd(self._get_cwd()):
+            agent = builder.build()
+            agent.context.shared_state["dialogues"] = BuyerDialogues(
+                agent.context.address
+            )
```
Comment (DavidMinarsch) [inline tests/test_fipa_end2end.py:292]: Instantiation is responsibility of the skill, not here.

Comment (solarw) [reply tests/test_fipa_end2end.py:292]: skill is creted programmatically and has no access to agent context till agent created by builder

Comment (solarw) [reply tests/test_fipa_end2end.py:292]: updated

File: tests/test_fipa_end2end.py (hunk 1 above)
Comment (solarw) [reply tests/test_fipa_end2end.py:114]: updated

File: tests/test_fipa_end2end.py (hunk 2 above)
Comment (solarw) [reply tests/test_fipa_end2end.py:146]: updated

File: tests/test_fipa_end2end.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+
+        seller_dialogue = self.dialogues.update(message)
+        if not seller_dialogue:
+            return
+        proposal_msg = seller_dialogue.reply(
+            target_message=message,
+            performative=FipaMessage.Performative.PROPOSE,
+            proposal=Description({"foo1": 1, "bar1": 2}),
+        )
+        self.context.outbox.put_message(proposal_msg)
```
Comment (DavidMinarsch) [inline tests/test_fipa_end2end.py:178]: Ideally we have the full fipa flow, so `cfp > propose > accept > match accept > inform`. But that can be done later and copy pasted from one of the skills.

File: tests/test_fipa_end2end.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    @classmethod
+    def stop(cls):
+        """Stop agent and tear down."""
+        cls.agent.stop()
+        cls.thread.join(20)
+        cls.teardown_class()
+
+
+class TestFipaEnd2End(Base):
+    """Test that echo skill works."""
```
Comment (marcofavorito) [inline tests/test_fipa_end2end.py:261]:
```suggestion
    """Test that the FIPA end-to-end demo works."""
```

File: tests/test_fipa_end2end.py [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+    def test_run(self):
+        """Run the echo skill sequence."""
```
Comment (marcofavorito) [inline tests/test_fipa_end2end.py:264]:
```suggestion
        """Run the FIPA end-to-end demo."""
```

File: tests/test_fipa_end2end.py [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        """Handle an evelope."""
+        dialogues = cast(FipaDialogues, self.context.dialogues)
+        buyer_dialogue = dialogues.update(message)
+        if not buyer_dialogue:
+            return
+        self.got_proposal = True
+
+
+class SellerHandler(Handler):
+    """Simple behaviour to count how many acts were called."""
```
Comment (marcofavorito) [inline tests/test_fipa_end2end.py:154]:
```suggestion
    """This class implement a simple seller handler for FIPA negotiations."""
```

Comment (marcofavorito) [reply tests/test_fipa_end2end.py:154]: or as above
```suggestion
    """Test SellerHandler."""
```

**Review by marcofavorito — APPROVED** [review]
LGTM

File: libs/go/aea_end2end/main.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		log.Fatal("Failed to start agent", err)
+	}
+	log.Print("successfully started AEA!")
+	
+	
+	dialogues := makeSellerDialogues(agent.Address())
+	envelope := agent.Get()
+    log.Printf("got incoming envelope: %s", envelope)
+    dialogues.Update(envelope)
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/main.go:83]: Then you should continue the dialogue: respond with PROPOSE, receive ACCEPT, send MATCH ACCEPT, etc

File: packages/fetchai/skills/fipa_dummy_buyer/README.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+# Fipa Dummy Buyer
+
+## Description
+
+A sample skill to interct with simple fipa seller.
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/fipa_dummy_buyer/README.md:5]:
```suggestion
A sample skill to interact with a simple seller using the `fetchai/fipa` protocol.
```

File: packages/fetchai/skills/fipa_dummy_buyer/dialogues.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+from aea.protocols.dialogue.base import Dialogue as BaseDialogue
+from aea.protocols.dialogue.base import DialogueLabel as BaseDialogueLabel
+from aea.skills.base import Model
+
+from packages.fetchai.protocols.fipa.dialogues import FipaDialogue as BaseFipaDialogue
+from packages.fetchai.protocols.fipa.dialogues import FipaDialogues as BaseFipaDialogues
+from packages.fetchai.protocols.fipa.message import FipaMessage
+
+
+class FipaDialogue(BaseFipaDialogue):
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/fipa_dummy_buyer/dialogues.py:42]: just do `FipaDialogue = BaseFipaDialogue` as you're not using Terms!

File: packages/fetchai/skills/fipa_dummy_buyer/behaviours.py [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+class FIPAInitializerBehaviour(TickerBehaviour):
+    """Fipa buyer cfp initializer."""
+
+    def __init__(self, **kwargs: Any) -> None:
+        """Init fipa behaviour."""
+        if "opponent_address" not in kwargs:
+            raise ValueError("Opponent address has to be specified for behaviour!")
+        self.opponent_address: str = cast(str, kwargs.pop("opponent_address", None))
+        self.is_enabled: bool = True
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/fipa_dummy_buyer/behaviours.py:38]: It's not disabled anywhere? We have a one-shot behaviour which might be more suitable here.

File: libs/go/aea_end2end/fetchai_private_key.txt [hunk 12]
```diff
@@ -0,0 +1 @@
+6f975c0d18150ca8af593abd721c094d1ed07b58a5ed01eba448c58adcb56b72
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/fetchai_private_key.txt:1]: Can we not generate this on the fly each time?

**Review by DavidMinarsch — COMMENTED** [review]
Please add a description in libs/go/aea_end2end/README.md how to run the end to end example from two terminals.

File: packages/fetchai/skills/fipa_dummy_buyer/behaviours.py (hunk 11 above)
Comment (solarw) [reply packages/fetchai/skills/fipa_dummy_buyer/behaviours.py:38]: it's disabled in handler on the propose received, the reason: agent do not perform discover so no clue when seller will be available

File: libs/go/aea_end2end/fetchai_private_key.txt (hunk 12 above)
Comment (solarw) [reply libs/go/aea_end2end/fetchai_private_key.txt:1]: probably added by mistake

File: packages/fetchai/skills/fipa_dummy_buyer/dialogues.py (hunk 10 above)
Comment (solarw) [reply packages/fetchai/skills/fipa_dummy_buyer/dialogues.py:42]: done

File: libs/go/aea_end2end/README.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
+# End2end example Go <> Python
+
+## Run example
+
+ensure all dependencies are installed: python, aea, golang
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/README.md:5]:
```suggestion
Ensure all dependencies are installed: python, aea and golang.
```

File: libs/go/aea_end2end/README.md [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+run to launch buyer agent
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/README.md:7]:
```suggestion
To launch the buyer agent run:
```

File: libs/go/aea_end2end/README.md [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+`./run_buyer.sh`
+
+run in another terminal to launch seller agent
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/README.md:10]:
```suggestion
In another terminal, to launch the seller agent run:
```

File: libs/go/aea_end2end/README.md [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+run to launch buyer agent
+`./run_buyer.sh`
+
+run in another terminal to launch seller agent 
+`./run_seller.sh`
+
+after a while both agents get connected to p2p fetchai network,
+perform fipa message exchange and show `FIPA INTERACTION COMPLETE` in output logs
+
+terminate every agent with `ctrl+c`
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/README.md:16]:
```suggestion
After a while both agents get connected to the ACN, perform a `fetchai/fipa` message exchange and show `FIPA INTERACTION COMPLETE` in output logs.

Terminate every agent with `ctrl+c`.
```

File: libs/go/aea_end2end/README.md [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+## Generate protocol
+
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/README.md:19]:
```suggestion

To generate a protocol, use the following approach:

```

File: libs/go/aea_end2end/README.md [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+``` bash
+aea create temp_agent
+cd temp_agent
+aea generate protocol -l PATH_TO_SPEC
+```
```
Comment (DavidMinarsch) [inline libs/go/aea_end2end/README.md:24]:
```suggestion
```

Provide additionally `--l go` to generate the go files.
```

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**